pstoreds/addr_book.go

```diff
@@ -2,398 +2,608 @@ package pstoreds
 	}
+}
 
-	mgr.cache.Remove(p)
-	// Attempt transactional KV deletion.
-	for i := 0; i < mgr.writeRetries; i++ {
-		if err = mgr.dbDelete(keys); err == nil {
-			break
+// loadRecord is a read-through fetch. It fetches a record from cache, falling back to the
+// datastore upon a miss, and returning a newly initialized record if the peer doesn't exist.
+//
+// loadRecord calls Refresh() on the record before returning it. If the record changes
+// as a result and `update=true`, an async flush is scheduled.
+//
+// If `cache=true`, the record is inserted in the cache when loaded from the datastore.
+func (ab *dsAddrBook) loadRecord(id peer.ID, cache bool, update bool) (pr *addrsRecord, err error) {
+	if e, ok := ab.cache.Get(id); ok {
+		pr = e.(*addrsRecord)
+		if pr.Refresh() && update {
+			ab.asyncFlush(pr)
 		}
-		log.Errorf("failed to delete addresses for peer %s: %s\n", p.Pretty(), err)
+		return pr, nil
 	}
 
+	txn, err := ab.ds.NewTransaction(true)
 	if err != nil {
-		log.Errorf("failed to avoid write conflict for peer %s after %d retries: %v\n", p.Pretty(), mgr.writeRetries, err)
-		return err
+		return nil, err
 	}
+	defer txn.Discard()
 
-	return nil
-}
+	key := addrBookBase.ChildString(b32.RawStdEncoding.EncodeToString([]byte(id)))
+	data, err := txn.Get(key)
 
-func (mgr *dsAddrBook) setAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration, mode ttlWriteMode) error {
-	// Keys and cleaned up addresses.
-	keys, addrs, err := keysAndAddrs(p, addrs)
-	if err != nil {
-		return err
+	if err != nil && err != ds.ErrNotFound {
+		return nil, err
 	}
 
-	mgr.cache.Remove(p)
-	// Attempt transactional KV insertion.
-	var existed []bool
-	for i := 0; i < mgr.writeRetries; i++ {
-		if existed, err = mgr.dbInsert(keys, addrs, ttl, mode); err == nil {
-			break
+	if err == nil {
+		pr = arPool.Get().(*addrsRecord)
+		if err = pr.Unmarshal(data); err != nil {
+			return nil, err
+		}
+		if pr.Refresh() && update {
+			ab.asyncFlush(pr)
 		}
-		log.Errorf("failed to write addresses for peer %s: %s\n", p.Pretty(), err)
+	} else {
+		pr = arPool.Get().(*addrsRecord)
+		pr.Id = &pb.ProtoPeerID{ID: id}
 	}
 
-	if err != nil {
-		log.Errorf("failed to avoid write conflict for peer %s after %d retries: %v\n", p.Pretty(), mgr.writeRetries, err)
-		return err
+	if cache {
+		ab.cache.Add(id, pr)
 	}
+	return pr, nil
+}
+
+// background runs the housekeeping process that takes care of:
+//
+// * GCing expired addresses from the datastore at regular intervals.
+// * persisting asynchronous flushes to the datastore.
+func (ab *dsAddrBook) background() {
+	// placeholder tickers.
+	pruneTimer, lookaheadTimer := new(time.Ticker), new(time.Ticker)
+
+	// populate the tickers after the initial delay has passed.
+	go func() {
+		select {
+		case <-time.After(ab.opts.GCInitialDelay):
+			pruneTimer = time.NewTicker(ab.opts.GCPurgeInterval)
+			lookaheadTimer = time.NewTicker(ab.opts.GCLookaheadInterval)
```

Comment:
This is racy.

Reply:
Fixed. I also separated flushing and GC to separate goroutines.